### ISSUE TITLE
Feature request/brainstorming: Drawing tablet support (wacom)

### ISSUE THREAD

**Issue by leviport** [pop-os/cosmic-settings#141]
This was mentioned in https://github.com/pop-os/cosmic-settings/issues/8, but I wanted to break this out specifically so as to not overwhelm that design thread with something that's not planned for the alpha. 

I figured that since I had my arm tied to one of these things for a decade while working my way through college in graphic design, I could share my thoughts about configurable features, current Gnome support, and what support in COSMIC DE might look like. There are many brands of tablets, but Wacom is definitely the big dog, so starting with Wacom support seems sensible to me.

### Features:
(Listed, in my estimation, from most important to least important.)

- __Mapping tablet to displays:__
Need to be able to map the tablet to a specific display. In my experience, it's not typical for multi-display setups to have the tablet mapped to the whole array of displays. 

- __Letterboxing / force aspect ratio mapping to display(s):__
The aspect ratio of a display (or group of displays) rarely matches the aspect ratio of a tablet. An ultrawide, for example, would have much faster cursor movement from side to side, than it would top to bottom. If I force letterbox with this display, a region of the tablet will be unmapped, and only a region with an aspect ratio that matches that of the display will be used.
_* Note: For some reason, letterboxing is not allowed in Gnome unless the tablet is mapped to one specific display. Mapping to a single display is the most common use case anyway, so this limitation probably doesn't affect very many users._

- __Left handed mode:__
Many tablets have their buttons on one side of the drawing area. For lefties, that means their hand would rest on them while they draw. Left handed mode rotates the tablet's mapping by 180° so that the physical tablet can be rotated, putting the buttons on the right side instead of on the left.

- __Pen pressure feel:__
Usually a slider from "soft" to "firm". Pens with "erasers" on the back usually have a second slide for the eraser as well.

- __Stylus button mapping:__
Most styli have 2 buttons on them. Since left click is done by touching the stylus to the tablet, usually one of these is right click by default, and the second is frequently customized based off the user's workflow. 3D workflows often change it to middle click, and a digital painter may want it bound to Ctrl + Z, or a hotkey that summons a quick access menu for a particular application. As a stretch component to this item, some configurators support multiple profiles for individual applications. So if I want certain bindings in Blender and completely different bindings in Krita, the application I currently have focused is detected, and the bindings are adjusted accordingly.

- __Tablet button mapping:__
These tablets usually have buttons, sliders, or dials, which the user can remap to send certain commands or key combos. The dial either repeats a key combo, or is bound to a scroll, zoom, or rotate action. When properly set up, the user won't have to reach for their keyboard, which is otherwise competing with the tablet for desk space.

- __Tracking mode:__
There are two different tracking modes. Absolute/tablet tracking makes the position of the cursor on the screen correspond with the position of the pen on the tablet. If the pen is picked up and moved to a different location, the cursor snaps to that location. Relative/mouse tracking does not snap the cursor to a new position if the pen is picked up and moved, but movements do move the cursor, much like how picking up up an optical mouse off the desk and setting it in a different location on the desk should not change the cursor's position. In my opinion, relative tracking is simply and objectively the wrong way to use a tablet, but some people probably prefer it for some reason.

- __Test your settings:__
Most of the configuration tools I've seen have a small rectangular area that the user can draw black lines so they can test their pressure settings.

- __Move and resize the mapped region on the tablet:__
This is one that Gnome Control Center does not have, but one I remember from Wacom's customizer from Windows. The active region on a tablet should be resizeable and moved to any position on the drawing area of the tablet. This somewhat ties into letterboxing, but it's much less important, so I made it into its own item down here.

- __LED settings:__
These tablets sometimes have customizable LEDs on them. The Intuos4 tablets even have little OLED displays next to the buttons. Being able to configure LEDs and mini OLED displays would be a very nice feature, but definitely a low priority.

**Comment by leviport** [comment]
## A tour of the Wacom configuration utility in Gnome Control Center

The "Wacom Tablet" panel does not appear unless a Wacom tablet is detected. Once it is, it looks like this:
![stylus](https://github.com/pop-os/cosmic-settings/assets/13512097/3699eb75-fa80-4288-a76c-a70e9abdad67)

The "stylus" and "tablet" sections are separate from each other, but I'm not entirely sure they need to be. 

The Stylus section is pretty simple. You can remap each button, and change the pressure of the nib and of the eraser (if present). It would be nice to be able to map custom key combos to the stylus buttons as well, but that functionality is not present:
![stylus-buttons](https://github.com/pop-os/cosmic-settings/assets/13512097/37adfbd5-04c8-4f1b-bd76-66ba09ab7913)

The "Test Your Settings" button in the top bar shows this little drawing area that I can use to test my settings:
![test-your-settings](https://github.com/pop-os/cosmic-settings/assets/13512097/ed73bb90-6c27-49d0-840e-624d67941d3b)

The Tablet section looks like this:
![tablet](https://github.com/pop-os/cosmic-settings/assets/13512097/64b0dec2-c36b-4818-8905-23a35c4ec39e)

Tracking mode can be set to either "Tablet (absolute)" which is correct, or the inferior "Touchpad (relative)" mode.
![tracking-mode](https://github.com/pop-os/cosmic-settings/assets/13512097/1cd6ead1-4ebb-4b5f-9aac-760532e93a52)

Left handed mode is a simple toggle.

I had trouble screenshotting the Map Buttons... dialog, so here is a cell phone picture of my screen:
![IMG_20240207_144708_014](https://github.com/pop-os/cosmic-settings/assets/13512097/cdcd4ff2-3790-466f-a4a6-eac484f85ef0)

The layout corresponds to the physical layout of the tablet:
![IMG_20240207_113202_213](https://github.com/pop-os/cosmic-settings/assets/13512097/3915284b-cd51-4d97-8841-da59f649986b)

In this dialog, you press one of the buttons or spin the dial on your tablet, then it allows you to customize it:
![IMG_20240207_144319_869](https://github.com/pop-os/cosmic-settings/assets/13512097/007c1be4-f7fa-4081-a0ec-602985588610)

And finally, the Map to Monitor dialog:
![display-mapping](https://github.com/pop-os/cosmic-settings/assets/13512097/333baa6f-713b-4596-80c0-803fd46d2498)

As mentioned before, the "Keep aspect ratio (letterbox)" switch becomes inactive if the tablet is not mapped to a single monitor.

**Comment by billksun** [comment]
> In my opinion, relative tracking is simply and objectively the wrong way to use a tablet, but some people probably prefer it for some reason.

It's perhaps the wrong way to draw with a tablet, but relative tracking mode is definitely a requirement be able to use a tablet as a general purpose pointing device. I have been suffering from wrist pains for years now, and have just recently purchased Wacom Intuos Pro S as a replacement for my mouse that's causing much of the pain. So, please hear me out on this different use case.

I am software engineer and also a gamer, and while absolute mode is just fine for desktop usage, gaming is a no-go. There's no way around it. In my research for a mouse replacement for both gaming and coding, I found that a drawing tablet not only can alleviate wrist pain, but also has the potential to have even more precision than a gaming mouse. However, in games like FPS or any game that uses the pointing device to pan or turn the viewport, this is can only be done in relative mode.

Relative mode might also work better with multi-display setups as a general mouse replacement. You can have displays of different sizes/resolutions/orientations and you don't have to worry about how that gets mapped to the tablet, possibly creating an awkwardly shaped active area, or if you have a smaller tablet, forced to use tablet in a higher sensitivity than desired.

Another configuration option that I feel should go with the relative mode implementation is the hover/proximity distance of the pen. You will be picking up the pen a lot when "mousing" around using relative mode, and being able to configure a comfortable lift height is absolutely crucial to long-term usage comfort.

A bonus option that would help with general desktop usage is having more options in the orientation of the tablet, similar to what [OpenTabletDriver](https://github.com/OpenTabletDriver/OpenTabletDriver?tab=readme-ov-file)'s "rotation" feature. Although I don't know how useful it is to have such fine grained rotation control, but I can definitely see the usefulness of having 90 degree increments, allowing you to place the tablet in landscape or portrait orientations. This gives users more options to optimally place their tablets amongst the other peripherals on their desktops.

**Comment by leviport** [comment]
I was being a bit cheeky with my remarks about relative mode. If I was seriously saying it shouldn't be included, I wouldn't have listed it in my feature list. Although I will say that I find the notion of gaming with a drawing tablet to be highly amusing. I might need to try it some time, just to see what it's like.

>Another configuration option that I feel should go with the relative mode implementation is the hover/proximity distance of the pen.

I'm not sure this is something that is configurable in any tablet settings applications. I think EMR tablets can only sense whether the pen is present or not. If I'm wrong and it does receive information about the distance of the pen, I do think that would be a good setting to include.

**Comment by billksun** [comment]
In regards to hover height, according to this screen capture of the Wacom's settings software in Windows from a KDE Plasma PR [comment](https://invent.kde.org/plasma/plasma-desktop/-/merge_requests/1963#note_845907), this is configurable (the "mouse height" setting, when "mouse mode" is selected):
![image](https://github.com/pop-os/cosmic-settings/assets/15665963/c7d3aaf9-8c33-46f5-baa0-7c6246311546)

From COSMIC's perspective, this will likely be configured through `libinput`'s [proximity events](https://wayland.freedesktop.org/libinput/doc/latest/tablet-support.html#handling-of-proximity-events). OpenTabletDriver implements this via the [Hover_Distance_Limiter](https://github.com/Kuuuube/Hover_Distance_Limiter) plugin.

Sorry for over-reacting about relative mode. I guess it's just that important for me, because the alternative would be to give up some of my favorite game genres, which I really don't want to :sweat:.

I currently use OpenTabletDriver just for my gaming needs and switch back to KDE Plasma's default tablet support for regular desktop use. This is because OpenTabletDriver does not handle any touch events from tablets, but I use my Wacom tablet as a trackpad when not gaming (multi-touch gestures really does improve desktop productivity!). I'm hoping COSMIC's solution would make this back-and-forth switch obsolete!

That being said, OpenTabletDriver has an option that I find very useful for gaming that I neglected to mention in my previous comment. In OpenTabletDriver, you can set a "Tip Threshold" for the left mouse button activation:
![image](https://github.com/pop-os/cosmic-settings/assets/15665963/f7ea4fad-886a-46a5-8fbf-8fc7e427b156)
This allows my pen tip to be in contact with the pad while not triggering a click event until I press harder into the pad. I find this to improve my precision in shooting games because of the slight friction when I drag the pen across the pad, and it minimizes the "shake" when pressing down on the pad. Without this, I would have to be hovering before pressing down, and I find it hard to maintain the pointer position while doing that.

**Comment by kekkoudesu** [comment]
This is a great overview of desirable features for a Drawing Tablet settings panel! I would definitely be impressed by it.

Just for the sake of comparison, here's KDE Plasma's Graphics Tablet settings: https://www.davidrevoy.com/data/images/blog/2024/debian-kde/2024-05-30_02_Tablet-Wayland-versus-X11.jpg
![image](https://github.com/user-attachments/assets/60826b4c-09f2-4675-b410-cf22edb1978b)

(It's a part of [this article](https://www.davidrevoy.com/article1030/debian-12-kde-plasma-2024-install-guide))

**Comment by Karvie0** [comment]
> AFAIK wacom tablets work OOB in Fedora and Wayland with libinput, I'm not sure what needs to be done to make the pen work. Any advice would be appreciated.

`libinput` is a library that lives inside the wayland compositor. Other areas of the compositor still have to handle all the information from `libinput` to send it to applications. There's also a lot of info from `libinput` that isn't meant to be sent to applications, but is instead meant to inform the compositor about how some things should be done (e.g., button mapping info), as well as some functions that allow the compositor to offload some transformations to `libinput`, such as for left-handed mode, or for custom pressure ranges.

TLDR; `libinput` provides hardware input info that the compositor still has to handle.

For now, it looks like you may be able to use OpenTabletDriver, as mentiond in https://github.com/pop-os/cosmic-comp/issues/313

**Comment by senntore** [comment]
Will this be implemented in Alpha 2?

**Comment by leviport** [comment]
No. New alphas are basically just spinning and releasing new ISOs. If you're on a 24.04 install (or 22.04 with COSMIC DE installed), the latest updates are released daily, and you will be receiving them as soon as they are released.

**Comment by SleepingCrows** [comment]
> For now, it looks like you may be able to use OpenTabletDriver, as mentiond in

Unfortunately OTD isn't a silver bullet for all tablets, Huion tablet support has a lot of gaps and work to do, and the first party drivers already have spotty Wayland support as is.

**Comment by arkonbob** [comment]
> > For now, it looks like you may be able to use OpenTabletDriver, as mentiond in
> 
> Unfortunately OTD isn't a silver bullet for all tablets, Huion tablet support has a lot of gaps and work to do, and the first party drivers already have spotty Wayland support as is.

Yes - for myself (Wacom Intuos Pro M pth-651) the situation has improved since I last looked, touch is now supported although the wheel remains broken (tbh I never use the wheel anyway). For others however, there are still a lot of gaps in support:

 https://opentabletdriver.net/Tablets

**Comment by billksun** [comment]
> touch is now supported

@arkonbob, how did you get touch to work?

**Comment by arkonbob** [comment]
@billksun I mean that touch is now supported for my tablet in Open Tablet Driver. I haven't confirmed this myself, I'm just going by the information in the repo (see the link I posted). Previously touch support was listed as broken there for the Intuos Pro M.

**Comment by senntore** [comment]
I don't want to rely on opentabletdriver even for Wacom tablets, as it is very clunky to use. If I enable pen pressure using linux artist mode then certain buttons in the pen would not work like their default behavior. Which makes it unusable to me.
I hope we can something similar to gnome's implementation as it works great out of the box.

**Comment by senntore** [comment]
Since this issue is being used to track tablet support, I would like to mention the issues related to currently present in COSMIC.
[ ] When we click on a window with the pen, the focus(& active hint) does not switch to that window. Because of this we cannot use keyboard shortcuts of the application used by pen unless we manually switch the focus using keyboard, mouse or touchpad.
[ ] Pen input does not work on any libcosmic related stuff like cosmic-panel, applibrary, launcher, dock & other cosmic applications. It works nicely on non-cosmic applications like Krita & Blender.

**Comment by BurntKrispe** [comment]
> If I enable pen pressure using linux artist mode then certain buttons in the pen would not work like their default behavior. 

How did you get artist mode to function? On my wacom bamboo and wacom one neither work when enabling artist mode. It's been necessary since running cosmic to switch to kde or gnome whenever I want to get any drawing done.

**Comment by BreeBreeByte** [comment]
I think it would be pretty helpful if there was a precision mode implemented. The wacom drivers for Windows has such a thing. The way it works is that by pressing a button, the user is able to 'zoom in' the drawing tablet region with a graphical representation of where it's zoomed in, onto the mouse cursor. I believe a plugin for opentabletdriver has this as well, although only the zooming and no graphical element.

**Comment by wwderw** [comment]
One thing that has kept me from moving to Cosmic is the lack of support for Wacom devices.  I'm looking thru this issue and it appears that this is more focused on the non display tablets.  Is this true or will there be support for the Cintiqs as well through this?

**Comment by leviport** [comment]
This issue is for drawing tablets, both with and without displays.

**Comment by Karvie0** [comment]
> One thing that has kept me from moving to Cosmic is the lack of support for Wacom devices. I'm looking thru this issue and it appears that this is more focused on the non display tablets. Is this true or will there be support for the Cintiqs as well through this?

The only real difference between display and non-display tablets is which monitor you're mapping the pen position onto. For this, we may want a calibration menu that lets the user hit targets with the pen. Any display related settings would likely go with the settings for the other displays

**Comment by wwderw** [comment]
> The only real difference between display and non-display tablets is which monitor you're mapping the pen position onto. For display tablets, we may want a calibration menu that lets the user hit targets with the pen. Any display related settings would likely go with the settings for the other displays

That mapping also extends to touch as well (depending on the Cintiq in question) and being able to turn that on/off independently.  Touch and pen inputs could also be mapped independently as well.  Can have one or the other extend beyond one screen (don't know why, but that is an option).  

Also the amount of express keys that are available as well.  I'm not counting the different "layers" available as most drivers on Linux only support one layer.  Which is fine if scripting is available.

Would there be support for scripting the mapping of the express keys like is available via xsetwacom on Xorg, or is it pretty much one and done via the GUI (while they can be remapped, that becomes tedious and not very efficient especially if there are different mappings for different use cases even within the same program, depending on workflow)?  KDE and Gnome don't have scripting on Wayland, so I don't know if that just is them not exposing that or it's not really a thing with Wayland.

**Comment by Karvie0** [comment]
> That mapping also extends to touch as well (depending on the Cintiq in question) and being able to turn that on/off independently.

Touch isn't something that's unique to drawing tablets, so I personally think it (and its calibration) should go in a different settings page. I can see there being a toggle for whether touch should be disabled based off pen proximity. Though it may instead be better to have that enabled as a default behavior and no toggle.

> Also the amount of express keys that are available as well. I'm not counting the different "layers" available as most drivers on Linux only support one layer. Which is fine if scripting is available.

I assume by "layers" you mean like keyboard layers. I could see that being useful, but in practice I'd suggest just using a keyboard instead if you really want to have a lot of flexibility. Split keyboards / macro-pads pair very nicely with display tablets.

Having the ability to bind a script to a tablet button is something that I've never thought of. Is this something people find genuinely useful, or is it mostly a workaround for lacking features in other areas? I'd like to hear people's thoughts on this.

**Comment by wwderw** [comment]
> 
> Touch isn't something that's unique to drawing tablets, so I personally think it (and its calibration) should go in a different settings page. I can see there being a toggle for whether touch should be disabled based off pen proximity. Though it may instead be better to have that enabled as a default behavior and no toggle.

Having a global on/off for touch, may not work if there is not just one touch monitor and I want to have specifically affect the Wacom.  Having pen proximity is no bueno if one accidentally lifts that pen, but still has their palm on there or another finger that happens to trigger something. I would personally have it with Wacom if there is an attached Wacom device with touch.
> 


> I assume by "layers" you mean like keyboard layers. I could see that being useful, but in practice I'd suggest just using a keyboard instead if you really want to have a lot of flexibility. Split keyboards / macro-pads pair very nicely with display tablets.

My Cintiq is off to the side, so using a keyboard is not an ergonomic option.  Nor would it be if it was inline with the keyboard as it would block the drawer to slide out that houses my keyboard)  I have used my streamdeck to compensate for this, and that would be a better option for how my computer is setup compared to just using the keyboard.  I do use split keyboards, but again, due to my layout, it is not ergonomic for such.  It would be better to just use the streamdeck at that point.
 


> Having the ability to bind a script to a tablet button is something that I've never thought of. Is this something people find genuinely useful, or is it mostly a workaround for lacking features in other areas? I'd like to hear people's thoughts on this.

I would argue that is a useful feature, because of a lack of options.  And given that it seems like Wacom support has to be exposed by the DE (from what I gather) this at least the ability of the individual user to work with their niche use and not have to worry the devs over such things.

Given the limitations of the GUI of Xsetwacom and Wayland (on KDE/GNOME) it is very useful, I just don't thinking many artists think about it, not like they would with python scripting in Blender or Krita.  Being limited to one layer on the express keys on Linux, that means that I have to either bind common actions to all the programs that I use (or the different workflows that I use for even one program) or I just stick the most important workflow on there.

I would have different bindings for Blender as a video editor, modeling/sculpting and for my 2D comics and/or animations when I use the grease pencil.  All of those workflows would have enough for me to fill multiple layers, but at least with scripting (and I actually had it setup to change depending on which program was in focus), I'm able to switch without having to think about it at all.  It was more inline with how the Windows/Mac OEM driver works for Wacom.

**Comment by rafaelcga** [comment]
Is this something that still doesn't work in general distros? I am using NixOS and am not able to get my XP-Pen tablet to work in any pressure detection mode in Krita, only in mouse mode. I am not sure if it is an idiosyncrasy of how COSMIC is installed in Nix or if other people are experiencing this too.

**Comment by aekobear** [comment]
> Is this something that still doesn't work in general distros? I am using NixOS and am not able to get my XP-Pen tablet to work in any pressure detection mode in Krita, only in mouse mode. I am not sure if it is an idiosyncrasy of how COSMIC is installed in Nix or if other people are experiencing this too.

Yes, it's still not working, tablet support is scheduled for cosmic epoch 2. Open Tablet Driver is also not working properly with cosmic / wayland in my experience. I've been forced to use KDE Plasma for any tablet-related work

**Comment by esplinr** [comment]
My tablet works out-of-the-box, including the buttons on my pen. But it took me a bit to figure out how to map it to a specific display.

Find out the tablet name with: `sudo libinput list-devices`
(install libinput from libinput-tools)
Find out the display name with `wlr-randr`

Then put this file in ~/.config/cosmic/com.system76.CosmicComp/v1/input_devices
```
/*
Map the graphics tablet to a specific monitor.
*/
{
 "UC-LOGIC Tablet WP8060U": (
    state: Enabled,
    map_to_output: Some("HDMI-A-3")
  )
}
```
Syntax errors will appear in journalctl as soon as you save because Cosmic immediately reads the settings.

For some reason, my tablet is upside down. I couldn't figure out how to rotate it, so I just physically flipped it over.

**Comment by BigJimsBigDinosaurFriends** [comment]
> My tablet works out-of-the-box, including the buttons on my pen.

Do you have pressure sensitivity? What tablet?

**Comment by lenhone** [comment]
> > My tablet works out-of-the-box, including the buttons on my pen.
> 
> Do you have pressure sensitivity? What tablet?

In my case pressure sensitivity works fine on an XPPen 15.6 Pro (V1) in Rnote.

As detailed above I installed  `libinput-tools` then ran `sudo libinput list-devices`  which gave: 
Device:                  UGTABLET 15.6 inch PenDisplay Stylus

Then created the below file with contents stated.
~/.config/cosmic/com.system76.CosmicComp/v1/input_devices

```
{
 "UGTABLET 15.6 inch PenDisplay Stylus": (
    state: Enabled,
    map_to_output: Some("HDMI-A-1")
  )
}
```

The stylus also follows my 180 deg flip I have set on HDMI-A-1.

**Comment by baaannnaaannn** [comment]
> I'm not a developer, but I can give my two cents as an artist. Both GNOME and Plasma have their shortcomings in this area, though GNOME is better than Plasma. Plasma's menu is too complicated, so people may be more inclined to ignore most of the settings, including ones they may have found useful if they were easier to understand. GNOME's settings are good, but they lean a bit too minimalistic IMO, and they don't need to be so separate.
> 
> I made a mockup of how the settings mentioned in the previous discussion could be presented nicely to the user. This isn't meant to be a suggestion for what the final menu should look like. This is just the clearest way to convey some ideas that I think are important to keep in mind.
>  
> ![Tablet Settings Mockup](https://github.com/user-attachments/assets/84fe1bdc-8872-444e-8f9d-152b94d74ef8)
> 
> There are a few main ideas here.
> - It'd be nice to have the buttons laid out in the menu as they are on the physical tablet, so you can click on the button you want to change. Same Idea for the pen side buttons, and the eraser.
> - It may also be nice to start editing a macro when you press the corresponding physical button on the tablet, so you don't have to click the menu.
> - Selecting an app at the top would let you set custom macros for that app.
> - The plus button to the right of the apps should show a list of open apps, or let you click on an open window to select an app.
> - Having a curve adjustment for pen pressure is the best way to go (based on personal experience).
> - The soft vs hard slider would be a faster way to adjust the pressure curve if you don't care about the detail so much.
> - I've put the pen settings higher than the display settings because they would be adjusted more often.
> 
> I'd also like to point out that the mockup isn't complete. It's missing some things, like a "default" option in the app select, pen height adjustment for mouse mode, and calibration for screen tablets. It may also be good to replace the "tip threshold" idea with `libinput`'s [custom pressure range](https://wayland.freedesktop.org/libinput/doc/latest/tablet-support.html#custom-tablet-tool-pressure-ranges). This could be presented as a slider with two control points.

I will use OS with this interface and colors
This is so cool

**Comment by harpetersen** [comment]
> > > My tablet works out-of-the-box, including the buttons on my pen.
> > 
> > 
> > Do you have pressure sensitivity? What tablet?
> 
> In my case pressure sensitivity works fine on an XPPen 15.6 Pro (V1) in Rnote.
> 
> As detailed above I installed `libinput-tools` then ran `sudo libinput list-devices` which gave: Device: UGTABLET 15.6 inch PenDisplay Stylus
> 
> Then created the below file with contents stated. ~/.config/cosmic/com.system76.CosmicComp/v1/input_devices
> 
> ```
> {
>  "UGTABLET 15.6 inch PenDisplay Stylus": (
>     state: Enabled,
>     map_to_output: Some("HDMI-A-1")
>   )
> }
> ```
> 
> The stylus also follows my 180 deg flip I have set on HDMI-A-1.

Worth pointing out that this method also worked for the Huion-Kamvas-25 Pro

**Comment by LuigiBrosNin** [comment]
> Worth pointing out that this method also worked for the Huion-Kamvas-25 Pro

I have a Huion Kamvas Pro 13, i would need the input from the pen recognized. this is what the workaround is for, correct?

**Comment by harpetersen** [comment]
> > Worth pointing out that this method also worked for the Huion-Kamvas-25 Pro
> 
> I have a Huion Kamvas Pro 13, i would need the input from the pen recognized. this is what the workaround is for, correct?

My Huion was working without it, but it was treating the pen as a mouse and not as a pen, meaning I wasn't getting things like pressure sensitivity. Getting the pen to work *at all* happened when I installed OpenTabletDriver. There was also a Huion add-on to OpenTabletDriver that I had to manually add that I forget where I got it (it was in the OpenTabletDriver forums, but I don't have a link), but I *think* it's included by default now?

**Comment by LuigiBrosNin** [comment]
> > > Worth pointing out that this method also worked for the Huion-Kamvas-25 Pro
> > 
> > 
> > I have a Huion Kamvas Pro 13, i would need the input from the pen recognized. this is what the workaround is for, correct?
> 
> My Huion was working without it, but it was treating the pen as a mouse and not as a pen, meaning I wasn't getting things like pressure sensitivity. Getting the pen to work _at all_ happened when I installed OpenTabletDriver. There was also a Huion add-on to OpenTabletDriver that I had to manually add that I forget where I got it (it was in the OpenTabletDriver forums, but I don't have a link), but I _think_ it's included by default now?

So all i have to do is configuring OpenTabletDriver? the last time i tried when i upgraded to COSMIC i had issues with it

Edit:
Aparently, this is how i should align my monitor according to OpenTabletDriver to be lined up with my display

<img width="1320" height="425" alt="Image" src="https://github.com/user-attachments/assets/f5dd2312-dbe2-40a5-b055-e410d5253cf4" />

Edit 2:
Pen input is not working, i have my pointer looking normal while it should be invisible and on Krita's brush preview stroke as showin in the pic (instead i just get my normal pointer)

<img width="448" height="428" alt="Image" src="https://github.com/user-attachments/assets/973dce93-60f3-45e3-a90e-202d3271e0a0" />

I have no idea how to fix this

**Comment by Karvie0** [comment]
i would like to suggest moving any future troubleshooting to discussions, instead of lengthening this issue

**Comment by ackered** [comment]
I tested a Wacom Bamboo CTH-460 on Pop!_OS 24.04 with COSMIC (Wayland). The kernel, libinput, and OpenTabletDriver all recognize the tablet correctly, but drawing applications (Krita, Blender) do not receive usable stylus input. The exact same installation works correctly when logging into GNOME, including pressure sensitivity. This suggests there may still be work to do in COSMIC's tablet input path in addition to the configuration UI discussed here.

OS:
Pop!_OS 24.04 LTS

Desktop:
COSMIC Wayland

Tablet:
Wacom Bamboo CTH-460 (USB ID: 056a:00d1)
Wacom Intuos 4 PTK-840

Applications tested:
- Krita (APT)
- Krita (Flatpak)
- Blender

Observed:
Touch input functions (only on Bamboo, but not the Intuos).
Stylus input is detected by the kernel and libinput but is not usable in applications.
Both the APT and Flatpak versions of Krita exhibited the same behavior under COSMIC.

Verified:
- wacom kernel driver loaded
- evtest reports ABS_X, ABS_Y, ABS_PRESSURE
- libinput debug-events reports TABLET_TOOL_* events
- udev tags device as ID_INPUT_TABLET=1
- OpenTabletDriver detects the tablet correctly

Control test:
Logging into GNOME on the same installation, using the same version of Krita, immediately restores pressure-sensitive pen input.

**Comment by DevWizard13** [comment]
Do we have a timeline for this feature? It's been in discussion for a really long time, and currently support still doesn't exist. I'm an artist and really need it. I'm honestly considering just switching to KDE or another more reliable desktop environment.

**Comment by LuigiBrosNin** [comment]
> Do we have a timeline for this feature? It's been in discussion for a really long time, and currently support still doesn't exist. I'm an artist and really need it. I'm honestly considering just switching to KDE or another more reliable desktop environment.

You can install Xfce alongside cosmic and go draw there since it's still using x11. i'll do the same, given it's one of the lightest stable DEs

**Comment by DevWizard13** [comment]
I just discovered sway seems to work largely out of the box when paired with opentabletdriver for drawing. Still, I really like cosmic and I really miss tablet support when I use it. I personally think this feature should be prioritized, as most other things that affect its fundamental usability for most workflows have been largely completed.

**Comment by leviport** [comment]
related: https://github.com/pop-os/cosmic-comp/issues/2721

Some support has been quietly added piece by piece since this issue was initially created. The main thing that is needed is a user-friendlier way to map a tablet to an output, but it still can be done by manually editing config files (see https://github.com/pop-os/cosmic-comp/issues/2687#issuecomment-5346387518 for details).

I will leave this feature request open since it relates more to the settings page itself.

**Comment by Tri4c3** [comment]
About manual configurations to files is very difficult if there is no tutorial out there.

Now the problem that time after time makes me not yet use my Wacom in popos is that the opentablet does not allow you to specify on which screen it will work. I have a monitor of 16 tablets and every time I move on my screen the laptop moves :\.

I hope that soon we can see this implementation with priority :/

**Comment by DevWizard13** [comment]
For me, it used to work, although the full virtual otd monitor was only mapped to the main monitor on my setup, monitor 1, which happened to be the smallest and worst monitor on my setup I usually don't even use (the built in laptop one). This was inconvenient because I prefer having it mapped onto the tablet, but it was workable. Recently, though, it started to glitch where it would randomly switch between the cursor being a cursor, brush, and hidden on Krita. This doesn't happen on other DEs, so it's cosmic specific. 

It's pretty simple, honestly, map the tablet surface to the entire collection of monitors instead of a specific monitor, then use otd to configure it further, at least until we can get a full implementation. In the meantime, I'll continue to use sway alongside cosmic until this is implemented since it does exactly this. I may even implement this myself, although DE dev is still an order of magnitude above my skill level.

**Comment by blakewilliams1** [comment]
@leviport Hey I've made a couple PRs to the cosmic-settings repo lately so it's pretty fresh in my mind how the pages are structured, if you'd like I could take a first pass at implementing the tablet settings page? As I understand it, the settings are stored in `~/.config/cosmic/com.system76.CosmicComp/v1/input_devices` as an [InputConfig](https://github.com/pop-os/cosmic-comp/blob/26daf75d93695d3ecd9418f4cfe8e0564ddc53db/cosmic-comp-config/src/input.rs#L11-L34)?

**Comment by leviport** [comment]
This will need a full UX design before it goes too far, but a PR might get the ball rolling if you want to work on it. It may need quite a few changes as the design is fleshed out though, so it's perfectly understandable if you'd rather hold off until the design is more complete.

**Comment by blakewilliams1** [comment]
hmm yeah, maybe just the bare minimum UI to get it working like what user YaPeNaBi did [over here](https://github.com/pop-os/cosmic-epoch/issues/3881#issuecomment-5749384769) to get it working? The entry they added to the input_devices only had `state` and `map_to_output` properties set. We could probably just show a checkbox for the state property and a dropdown to pick between displays. There's plenty of examples on dropdowns for picking displays in the "Displays" page to go off of.

**Comment by blakewilliams1** [comment]
@leviport hey I'm making good progress building a basic UI for those critical properties for tablet configs, but I had a couple questions about it to make sure it'll actually be useful until official design work is done:

1: The default config for a tablet should be a DeviceState property set to Disabled, and the map_to_output property not set to anything, right?
2: Right now when the user clicks into the tablet settings page (similar to how you click into keyboard or mouse settings pages) I load pre-existing configs present in `input_devices` config file and also scan for tablet input devices matching ID_INPUT_TABLET property. We should show the combination of both of these right? For example let's say I've been using a XP Pen tablet but it breaks and now I'm using a new Wacom. If I previously had a config for the XP Pen saved in `input_devices` and my Wacom is currently plugged in then I should see the old config entry for the XP Pen as well as default config settings for the Wacom, right? Alternatively we just show UI for only tablets currently plugged in and then update or add configs as the user interacts with the UI. 
3: If we show configs present in the `input_devices` representing tablets NOT currently plugged in, we should show a trash symbol button next to that entry to allow users to clear out stale configs? If yes then I'd imagine the delete button would only show on configs present in the config file but not currently plugged in. 

wdyt?
